src/fiber/fiber.ml

```diff
@@ -1,243 +1,455 @@
-     fiber that terminates when all the fiber in the sub-context have
-     terminated. *)
-  val wait_errors : (unit -> 'a t) -> ('a, unit) result t
+    val unset : 'a t -> (unit -> 'b fiber) -> 'b fiber
+  end
 
-  (* Set the current error handler. [on_error] is called in the current
-     execution context. *)
-  val set_error_handler :
-    on_error:(Exn_with_backtrace.t -> unit t) -> ('a -> 'b t) -> 'a -> 'b t
+  type 'a k
 
-  val vars : unit -> Univ_map.t
+  (* Suspend the execution of a fiber *)
+  val suspend : ('a k -> unit) -> 'a fiber
 
-  val set_vars : Univ_map.t -> ('a -> 'b t) -> 'a -> 'b t
+  (* Restart a suspended fiber. Must only be called once on a given suspended
+     fiber *)
+  val restart : 'a k -> 'a -> unit fiber
 
-  val run : 'a t -> iter:(unit -> unit) -> 'a
+  type fill = Fill : 'a Ivar.t * 'a -> fill
 
-  val reraise_all : Exn_with_backtrace.t list -> unit
+  val run : 'a t -> iter:(unit -> fill Nonempty_list.t) -> 'a
 end = struct
-  type t =
-    { on_error : Exn_with_backtrace.t k option
-          (* This handler must never raise *)
+  type 'a t = ('a -> effect) -> effect
+
+  and effect =
+    | Read_ivar : 'a ivar * ('a -> effect) -> effect
+    | Fill_ivar : 'a ivar * 'a * (unit -> effect) -> effect
+    | Suspend : ('a k -> unit) * ('a -> effect) -> effect
+    | Restart : 'a k * 'a * (unit -> effect) -> effect
+    | Get_var : 'a Univ_map.Key.t * ('a option -> effect) -> effect
+    | Set_var :
+        'a Univ_map.Key.t * 'a * (unit -> 'b t) * ('b -> effect)
+        -> effect
+    | Unset_var : 'a Univ_map.Key.t * (unit -> 'b t) * ('b -> effect) -> effect
+    | With_error_handler :
+        (unit -> 'a t) * (Exn_with_backtrace.t -> Nothing.t t) * ('a -> effect)
+        -> effect
+    | Unwind : context * ('a -> effect) * 'a -> effect
+    | Map_reduce_errors :
+        (module Monoid with type t = 'a)
+        * (Exn_with_backtrace.t -> 'a t)
+        * (unit -> 'b t)
+        * (('b, 'a) result -> effect)
+        -> effect
+    | End_of_fiber of unit
+    | Never of unit
+    (* Add a dummy unit argument to [End_of_fiber] and [Never] so that all
+       constructors are boxed, which removes a branch in the pattern match. *)
+    | End_of_map_reduce_error_handler : (_, _) on_release' -> effect
+    | Fork : (unit -> effect) * (unit -> effect) -> effect
+    | Parallel_iter : 'a list * ('a -> unit t) * (unit -> effect) -> effect
+    | Parallel_map : 'a list * ('a -> 'b t) * ('b list -> effect) -> effect
+    | Reraise_all : Exn_with_backtrace.t list -> effect
+    | Done of value
+
+  and 'a ivar = { mutable state : ('a, [ `Full | `Empty ]) ivar_state }
+
+  and ('a, _) ivar_state =
+    | Full : 'a -> ('a, [> `Full ]) ivar_state
+    | Empty : ('a, [> `Empty ]) ivar_state
+    | Empty_with_readers :
+        context * ('a -> effect) * ('a, [ `Empty ]) ivar_state
+        -> ('a, [> `Empty ]) ivar_state
+
+  and value = ..
```

Comment:
Could you add a comment for this mysterious type?

Reply:
Sure